src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -618,45 +599,59 @@ H5FD__subfiling_get_default_config(hid_t fapl_id, H5FD_subfiling_config_t *confi
     }
 
     /* Check if any MPI parameters were set on the FAPL */
-    if (H5Pget_mpi_params(fapl_id, &comm, &info) < 0)
-        H5_SUBFILING_GOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "can't get MPI Comm/Info");
+    if (NULL == (plist = (H5P_genplist_t *)H5I_object(fapl_id)))
+        HGOTO_ERROR(H5E_VFL, H5E_BADID, FAIL, "can't find object for ID");
+    if (H5P_get(plist, H5F_ACS_MPI_PARAMS_COMM_NAME, &comm) < 0)
+        HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "can't get MPI communicator from plist");
+    if (H5P_get(plist, H5F_ACS_MPI_PARAMS_INFO_NAME, &info) < 0)
+        HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "can't get MPI info from plist");
     if (comm == MPI_COMM_NULL) {
         comm = MPI_COMM_WORLD;
 
         /* Set MPI_COMM_WORLD on FAPL if no MPI parameters were set */
-        if (H5Pset_mpi_params(fapl_id, comm, info) < 0)
-            H5_SUBFILING_GOTO_ERROR(H5E_PLIST, H5E_CANTSET, FAIL, "can't set MPI Comm/Info");
+        if (H5P_set(plist, H5F_ACS_MPI_PARAMS_COMM_NAME, &comm) < 0)
+            HGOTO_ERROR(H5E_VFL, H5E_CANTSET, FAIL, "can't set MPI communicator");
+        if (H5P_set(plist, H5F_ACS_MPI_PARAMS_INFO_NAME, &info) < 0)
+            HGOTO_ERROR(H5E_VFL, H5E_CANTSET, FAIL, "can't set MPI info object");
     }
 
     /* Create a default FAPL and choose an appropriate underlying driver */
-    if ((config_out->ioc_fapl_id = H5Pcreate(H5P_FILE_ACCESS)) < 0)
-        H5_SUBFILING_GOTO_ERROR(H5E_PLIST, H5E_CANTCREATE, FAIL, "can't create default FAPL");
+    if (H5FD__copy_plist(H5P_FILE_ACCESS_DEFAULT, &config_out->ioc_fapl_id) < 0)
+        HGOTO_ERROR(H5E_VFL, H5E_CANTCREATE, FAIL, "can't create default FAPL");
+    if (NULL == (plist = (H5P_genplist_t *)H5I_object(config_out->ioc_fapl_id)))
+        HGOTO_ERROR(H5E_VFL, H5E_BADID, FAIL, "can't find object for ID");
 
     if (config_out->require_ioc) {
-        if (H5Pset_mpi_params(config_out->ioc_fapl_id, comm, info) < 0)
-            H5_SUBFILING_GOTO_ERROR(H5E_PLIST, H5E_CANTSET, FAIL, "can't get MPI Comm/Info on IOC FAPL");
+        H5FD_ioc_config_t ioc_config;
 
-        if (H5Pset_fapl_ioc(config_out->ioc_fapl_id, NULL) < 0)
-            H5_SUBFILING_GOTO_ERROR(H5E_PLIST, H5E_CANTSET, FAIL, "can't set IOC VFD on IOC FAPL");
+        if (H5P_set(plist, H5F_ACS_MPI_PARAMS_COMM_NAME, &comm) < 0)
+            HGOTO_ERROR(H5E_VFL, H5E_CANTSET, FAIL, "can't set MPI communicator");
+        if (H5P_set(plist, H5F_ACS_MPI_PARAMS_INFO_NAME, &info) < 0)
+            HGOTO_ERROR(H5E_VFL, H5E_CANTSET, FAIL, "can't set MPI info object");
+
+        if (H5FD__subfiling_get_default_ioc_config(&ioc_config) < 0)
```

Comment:
I think this highlights an architectural issue in that Subfiling needs to be able to set the underlying I/O concentrator VFD using a default config without being able to make a public API call here. Subfiling was meant to be a modular architecture, with the IOC VFD being a reference implementation of an I/O concentrator, but this tends to tie it right to the reference implementation. Fine for now, but I think a better solution is needed in general for the case of VFDs stacked on top of other VFDs.

Reply:
Agree - we should update the stackable VFDs RFC at some point, with the learnings from the VOL connectors.